### PR TITLE
Chain monitor

### PR DESCRIPTION
**Pull request by Solovyov1796**
This change is [Reviewable](https://reviewable.io/reviews/0glabs/0g-monitor/29)

### CODE REVIEW

File: blockchain/ethermint.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	return BlockInfo{
+		Height:    resBlockNumber,
+		Timestamp: resTimestamp,
+		Hash:      hsah,
+		TxHashes:  hashList,
+	}, nil
+}
+
+func EthFetchTxReceiptStatus(url string, txHash string) (uint64, error) {
+	reqBody := map[string]interface{}{
```
Comment (0g-wh) [inline blockchain/ethermint.go:67]: use getBlockReceipt if you need to check the status of all txns

File: cmd/blockchain/blockchain.go [hunk 2]
```diff
@@ -1,25 +1,30 @@
...
 import (
-	"sync"
-
 	"github.com/0glabs/0g-monitor/blockchain"
 	"github.com/0glabs/0g-monitor/utils"
-	"github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
 )
 
+// const (
```
Comment (0g-wh) [inline cmd/blockchain/blockchain.go:9]: delete comment lines if the code is not needed

File: blockchain/cometbft.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		return 0, err
+	}
+
+	return int(cnt), nil
+}
+
+func rpcGetBlockValidatorCnt(url string, height uint64) (int, error) {
+	client := resty.New()
+	var result map[string]interface{}
+	resp, err := client.R().SetResult(&result).Get(fmt.Sprintf("%s/validators?height=%d", url, height))
```
Comment (0g-wh) [inline blockchain/cometbft.go:46]: the default page size is 30, need to call multi times to get all validators

File: blockchain/serve.go [hunk 4]
```diff
@@ -43,43 +41,203 @@ func Monitor(config Config) {
...
 		v.Update(config.ValidatorReport)
+		if v.jailed {
+			jailedCnt++
+		}
+	}
+
+	activeValidatorCount := len(validators) - jailedCnt
+	metrics.GetOrRegisterHistogram(validatorActiveCountPattern).Update(int64(activeValidatorCount))
+	percentage := 100 * float64(activeValidatorCount) / float64(len(validators))
+	if percentage-float64(67) >= 0 {
```
Comment (0g-wh) [inline blockchain/serve.go:204]: activeValidatorCount is important but the percentage should be calculated based on voting power

Comment (Solovyov1796) [reply blockchain/serve.go:204]: the validators are 0g holding, not all validators in chain. so the percentage is about the active validators holing by 0g. And the voting power are calculated dynamically in real time based on the stake amount and the validator status. Validator jailed will trigger the calculate and then voting power will be changed.  the detail code in https://github.com/0glabs/cosmos-sdk/blob/0glabs-v0.47.10-release-x/x/staking/keeper/val_state_change.go#ApplyAndReturnValidatorSetUpdates

File: blockchain/cometbft.go (hunk 3 above)
Comment (Solovyov1796) [reply blockchain/cometbft.go:46]: peek signed validator count from field "total"

**Review by 0g-wh — APPROVED** [review]

**Pull request merged**